### PR TITLE
Add better performing versions for groupwise and depthwise convolutions

### PR DESCRIPTION
**Pull request by dskhudia**
Summary: Groupwise and depthwise convolutions become faster with this diff

Differential Revision: D16264144

### CODE REVIEW

**Comment by twmht** [comment]
@dskhudia 

Do you benchmark this PR?

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Please add a benchmark test for the kernels and add the benchmarking results before and after the change as part of documentation for this PR. Thanks!

File: torch/nn/quantized/modules/conv.py [hunk 1]
```diff
@@ -64,7 +64,7 @@ def weight(self):
 
     @weight.setter
     def weight(self, w):
-        self._packed_weight = torch.ops.quantized.fbgemm_conv_prepack(w, self.groups)
+        self._packed_weight = torch.ops.quantized.fbgemm_conv_prepack(w, self.stride, self.padding, self.dilation, self.groups)
```
Comment (jerryzh168) [inline torch/nn/quantized/modules/conv.py]: an unrelated question: do we have unpack for conv?

**Comment by dskhudia** [comment]
> @dskhudia
> 
> Did you benchmark this PR?

@twmht
I'll add some benchmarking results soon. BTW, this is for convolutions on int8 data rather than fp32.

**Comment by dskhudia** [comment]
> Please add a benchmark test for the kernels and add the benchmarking results before and after the change as part of documentation for this PR. Thanks!

Sure. I'll add benchmarking results in the summary. I have a followup PR that adds benchmarking. https://github.com/pytorch/pytorch/pull/22895

File: torch/nn/quantized/modules/conv.py (hunk 1 above)
Comment (dskhudia) [reply torch/nn/quantized/modules/conv.py]: @jerryzh168 
We made sure to add unpack for conv. Details in https://github.com/pytorch/FBGEMM/pull/103, https://github.com/pytorch/FBGEMM/pull/105, https://github.com/pytorch/FBGEMM/pull/106, https://github.com/pytorch/FBGEMM/commit/815139b1ba72ed43571084f58c7fdf4ca3c991d5, https://github.com/pytorch/FBGEMM/commit/64a2c73a425e4113839f2c2b596ea28d632d20f6

**Comment by dskhudia** [comment]
@raghuramank100 Please see summary for before and after results.

File: torch/nn/quantized/modules/conv.py (hunk 1 above)
Comment (jerryzh168) [reply torch/nn/quantized/modules/conv.py]: is this ready or still ongoing?

Comment (jerryzh168) [reply torch/nn/quantized/modules/conv.py]: oh I saw qconv_unpack, Thanks

Comment (dskhudia) [reply torch/nn/quantized/modules/conv.py]: This is ready. I might have to change your qat modules if they have landed.

File: aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp [hunk 2]
```diff
@@ -17,17 +17,50 @@ namespace {
...
-    int KDim_per_group = kernel_h * kernel_w * input_channels_per_group;
+
+    // mini-batch doesn't have any impact on how we pack weights
+    // so we pass it as 1
+    // Input image height/width also don't have any impact on how we pack
+    // weights so we can pass any values
+    fbgemm::conv_param_t<2> conv_p(
+        1, // Mini-Batch
+        input_channels_per_group * groups, // input channels
+        output_channels,
```
Comment (raghuramank100) [inline aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp]: Why do we need mini-batch and image size as parameters to this function if they do not have any effect? Can we remove these arguments?

Comment (dskhudia) [reply aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp]: The interface for packing takes struct conv_param_t as argument. This way we pass the same conv_param_t struct to both conv operation as well as packing. We will have to change the interface and define a new struct if we want to remove these parameters and that would just work for packing only. The other way of doing this is to add a function in fbgemm that would return conv_param_t and does't take mini-batch and imagesize as input parameters.

File: aten/src/ATen/native/quantized/cpu/qconv.cpp [hunk 3]
```diff
@@ -99,11 +99,10 @@ class QConv2dInt8 final : public c10::OperatorKernel {
     PackedConvWeight& pack_ptr =
         cpp_custom_type_hack::cast<PackedConvWeight>(packed_weight);
     auto packB = pack_ptr.w.get();
-    // packB->printPackedMatrix("PackedB inside QConv2dInt8:");
```
Comment (jianyuh) [inline aten/src/ATen/native/quantized/cpu/qconv.cpp]: Still keep this for debugging purpose? Do we have printPackedMatrix function in the general Conv interface?

Comment (dskhudia) [reply aten/src/ATen/native/quantized/cpu/qconv.cpp]: We don't. Let me remove this.

**Review by raghuramank100 — APPROVED** [review]

**Pull request closed without merging**